### PR TITLE
Apply sstable io error handler to exceptions generated when opening file

### LINKED ISSUE #19753: SSTable io_error_handler is not applied to file open
In the discussion on which exceptions should be treated as "recoverable", when the server should fall into "endless retry" + "isolate" mode, the fact that we treat IO errors generated by actual IO (read/write) differently from errors generated by file open. This is strange, since EPERM for example is one of the errors we typically normally treat as "keep retrying by isolate". 

By extension this also makes it questionable any attempt to fold more "high-level" exceptions into the above category (read: exceptions caused by extensions) - when should we isolate nodes and not?

Here is a not-so-small dual test case showing the difference in behaviour:

```

SEASTAR_TEST_CASE(test_exceptions_in_flush_on_sstable_write) {
    auto ext = std::make_shared<db::extensions>();
    auto cfg = seastar::make_shared<db::config>(ext);

    class myext : public sstables::file_io_extension {
    public:
        bool should_fail = false;
        bool did_fail = false;

        future<file> wrap_file(sstable& t, component_type type, file f, open_flags flags) override {
            if (should_fail) {
                class myimpl : public seastar::file_impl {
                    file _file;
                    myext& _myext;
                public:
                    myimpl(file f, myext& ext)
                        : _file(std::move(f))
                        , _myext(ext)
                    {}
                    void fail() const {
                        if (_myext.should_fail) {
                            _myext.did_fail = true;
                            BOOST_TEST_MESSAGE("Throwing exception");
                            throw std::system_error(EACCES, std::system_category());
                        }
                    }
                    future<size_t> write_dma(uint64_t pos, const void* buffer, size_t len, io_intent* intent) override {
                        fail();
                        return get_file_impl(_file)->write_dma(pos, buffer, len, intent);
                    }
                    future<size_t> write_dma(uint64_t pos, std::vector<iovec> iov, io_intent* intent) override {
                        fail();
                        return get_file_impl(_file)->write_dma(pos, std::move(iov), intent);
                    }
                    future<size_t> read_dma(uint64_t pos, void* buffer, size_t len, io_intent* intent) override {
                        fail();
                        return get_file_impl(_file)->read_dma(pos, buffer, len, intent);
                    }
                    future<size_t> read_dma(uint64_t pos, std::vector<iovec> iov, io_intent* intent) override {
                        fail();
                        return get_file_impl(_file)->read_dma(pos, std::move(iov), intent);
                    }
                    future<> flush(void) override {
                        fail();
                        return get_file_impl(_file)->flush();
                    }
                    future<struct stat> stat(void) override {
                        fail();
                        return get_file_impl(_file)->stat();
                    }
                    future<> truncate(uint64_t length) override {
                        fail();
                        return get_file_impl(_file)->truncate(length);
                    }
                    future<> discard(uint64_t offset, uint64_t length) override {
                        fail();
                        return get_file_impl(_file)->discard(offset, length);
                    }
                    future<> allocate(uint64_t position, uint64_t length) override {
                        fail();
                        return get_file_impl(_file)->allocate(position, length);
                    }
                    future<uint64_t> size(void) override {
                        fail();
                        return get_file_impl(_file)->size();
                    }
                    future<> close() override {
                        fail();
                        return get_file_impl(_file)->close();
                    }
                    std::unique_ptr<seastar::file_handle_impl> dup() override {
                        fail();
                        return get_file_impl(_file)->dup();
                    }
                    subscription<directory_entry> list_directory(std::function<future<> (directory_entry de)> next) override {
                        fail();
                        return get_file_impl(_file)->list_directory(std::move(next));
                    }
                    future<temporary_buffer<uint8_t>> dma_read_bulk(uint64_t offset, size_t range_size, io_intent* intent) override {
                        fail();
                        return get_file_impl(_file)->dma_read_bulk(offset, range_size, intent);
                    }
                };
                co_return file(make_shared<myimpl>(std::move(f), *this));
            }
            co_return f;
        }
    };
    auto mep = std::make_unique<myext>();
    auto& me = *mep;
    ext->add_sstable_file_io_extension("test", std::move(mep));

    co_await do_with_cql_env([&](cql_test_env& env) -> future<> {

        co_await env.execute_cql(fmt::format("create table t0 (pk text primary key, v text)"));

        me.should_fail = true;

        int i = 0;

        BOOST_TEST_MESSAGE("Wait for fail");

        auto f = make_ready_future<>();

        while (!me.did_fail) {
            std::string pk = "apa" + std::to_string(i++);
            std::string v = "ko";
            co_await env.execute_cql(fmt::format("insert into ks.t0 (pk, v) values ('{}', '{}')", pk, v));

            f = f.then([&] {
                return env.db().invoke_on_all([] (replica::database& db) {
                    return db.flush_all_memtables();
                });
            });
        }

        BOOST_REQUIRE(me.did_fail);
        BOOST_TEST_MESSAGE("Reset fail trigger");

        me.should_fail = false;

        co_await env.get_storage_service().invoke_on(0, [&](service::storage_service& ss) {
            BOOST_REQUIRE(ss.is_stopping()); // should have been isolated here.
        });

        BOOST_TEST_MESSAGE("Trying to stop");

        co_await std::move(f);
    }, cfg);
}

SEASTAR_TEST_CASE(test_exceptions_in_flush_on_sstable_open) {
    auto ext = std::make_shared<db::extensions>();
    auto cfg = seastar::make_shared<db::config>(ext);

    class myext : public sstables::file_io_extension {
    public:
        bool should_fail = false;
        bool did_fail = false;

        future<file> wrap_file(sstable& t, component_type type, file f, open_flags flags) override {
            if (should_fail) {
                did_fail = true;
                throw std::system_error(EACCES, std::system_category());
            }
            co_return f;
        }
    };
    auto mep = std::make_unique<myext>();
    auto& me = *mep;
    ext->add_sstable_file_io_extension("test", std::move(mep));

    co_await do_with_cql_env([&](cql_test_env& env) -> future<> {

        co_await env.execute_cql(fmt::format("create table t0 (pk text primary key, v text)"));

        me.should_fail = true;

        int i = 0;

        BOOST_TEST_MESSAGE("Wait for fail");

        auto f = make_ready_future<>();

        while (!me.did_fail) {
            std::string pk = "apa" + std::to_string(i++);
            std::string v = "ko";
            co_await env.execute_cql(fmt::format("insert into ks.t0 (pk, v) values ('{}', '{}')", pk, v));

            f = f.then([&] {
                return env.db().invoke_on_all([] (replica::database& db) {
                    return db.flush_all_memtables();
                });
            });
        }

        BOOST_REQUIRE(me.did_fail);
        BOOST_TEST_MESSAGE("Reset fail trigger");

        me.should_fail = false;

        co_await env.get_storage_service().invoke_on(0, [&](service::storage_service& ss) {
            BOOST_REQUIRE(ss.is_stopping()); // should have been isolated here.
        });

        BOOST_TEST_MESSAGE("Trying to stop");

        co_await std::move(f);
    }, cfg);
}

```

The second case will fail, because io_error_handler is not applied to the path that throws, and while we endlessly retry even here, we will not isolate the node.

### PR DESCRIPTION
**Pull request by elcallio**
Fixes #19753

SSTable file open provides an `io_error_handler` instance which is applied to a file-wrapper to process any IO errors happing during read/write via the handler in `storage_service`, which in turn will effectively disable the node. However, this is not applied to the actual open operation itself, i.e. any exception generated by the file open call itself will instead just escape to caller. 

This PR adds filtering via the `error_handler` to sstable open + makes `storage_service` "isolate" mechanism non-module-static (thus making it testable) and adds tests to check we exhibit the same behaviour in both cases. 

The main motivation for this issue it discussions that secondary level IO issues (i.e. caused by extensions) should trigger the same behaviour as, for example, running out of disk space.

### CODE REVIEW

File: sstables/sstables.cc [hunk 1]
```diff
@@ -142,8 +142,14 @@ future<file> sstable::new_sstable_component_file(const io_error_handler& error_h
...
 
-    return f.handle_exception([this, type] (auto ep) {
+    return f.handle_exception([this, type, &error_handler] (auto ep) {
         sstlog.error("Could not create SSTable component {}. Found exception: {}", filename(type), ep);
+        try {
+            error_handler(ep);
+        } catch (...) {
+            ep = std::current_exception();
+        }
+        std::rethrow_exception(ep);
```
Comment (denesb) [inline sstables/sstables.cc:152]: `std::move(ep)`

Comment (denesb) [reply sstables/sstables.cc:152]: Also, how is the above try-catch-rethrow different from just not catching whatever `error_handler()` throws?

File: service/storage_service.hh [hunk 2]
```diff
@@ -527,6 +527,7 @@ private:
     sharded<service::tablet_allocator>& _tablet_allocator;
     sharded<cdc::generation_service>& _cdc_gens;
     sharded<db::view::view_builder>& _view_builder;
+    std::atomic<bool> _isolated = false;
```
Comment (denesb) [inline service/storage_service.hh:530]: Does this still need to be atomic, now that it is a regular member of `storage_service`?

File: test/boost/memtable_test.cc [hunk 3]
```diff
@@ -1170,3 +1173,166 @@ SEASTAR_TEST_CASE(flushing_rate_is_reduced_if_compaction_doesnt_keep_up) {
...
+
+    co_await do_with_cql_env([&](cql_test_env& env) -> future<> {
+
+        co_await env.execute_cql(fmt::format("create table t0 (pk text primary key, v text)"));
+
+        should_fail = true;
+
+        int i = 0;
+
+        BOOST_TEST_MESSAGE("Wait for fail");
```
Comment (denesb) [inline test/boost/memtable_test.cc:1190]: suggestion: we have `testlog`, which provides a nicer end-user experience than `BOOST_TEST_MESSAGE`.

File: test/boost/memtable_test.cc [hunk 4]
```diff
@@ -1170,3 +1173,166 @@ SEASTAR_TEST_CASE(flushing_rate_is_reduced_if_compaction_doesnt_keep_up) {
...
+        bool isolated = false;
+        for (int i = 0; i < 10; ++i) {
+            isolated = co_await env.get_storage_service().invoke_on(0, [&](service::storage_service& ss) {
+                return ss.is_isolated(); 
+            });
+            if (isolated) {
+                break;
+            }
+            // isolation is not syncnronous;
+            co_await sleep(2s);
```
Comment (denesb) [inline test/boost/memtable_test.cc:1219]: We have the `eventually()` and `eventually_true()` helpers, which already implement something like this, see https://github.com/scylladb/scylladb/blob/master/test/lib/eventually.hh.

**Review by denesb — CHANGES_REQUESTED** [review]

File: sstables/sstables.cc (hunk 1 above)
Comment (elcallio) [reply sstables/sstables.cc:152]: It is not. But it is not supposed to be there. Sorry, copy-paste from playing with coroutinization. There is a return statement after...

File: service/storage_service.hh (hunk 2 above)
Comment (elcallio) [reply service/storage_service.hh:530]: No, it really does not.

File: test/boost/memtable_test.cc (hunk 3 above)
Comment (elcallio) [reply test/boost/memtable_test.cc:1190]: Should we not redirect one to the other? I personally tend to use the "-l" option on boost tests quite a bit to know where I am...

File: test/boost/memtable_test.cc (hunk 4 above)
Comment (elcallio) [reply test/boost/memtable_test.cc:1219]: Neat.

File: service/storage_service.hh (hunk 2 above)
Comment (elcallio) [reply service/storage_service.hh]: Changed

File: test/boost/memtable_test.cc (hunk 3 above)
Comment (elcallio) [reply test/boost/memtable_test.cc]: Changed

File: test/boost/memtable_test.cc (hunk 4 above)
Comment (elcallio) [reply test/boost/memtable_test.cc:1221]: Does not work here, because neither context in this case is a seastar thread

**Comment by elcallio** [comment]
Rebased and fixed comments above

**Comment by bhalevy** [comment]
Cc @lersek (re: dir_wrapper error handling)

**Comment by elcallio** [comment]
@annastuchlik - sorry, you got tagged by automation due to a rebase problem. Same with @nyh

File: test/boost/memtable_test.cc (hunk 3 above)
Comment (denesb) [reply test/boost/memtable_test.cc]: This is a matter of taste I suppose, that is why I prefixed my comment with "suggestion". I personally prefer the logger.

**Review by denesb — APPROVED** [review]

**Comment by lersek** [comment]
> Cc @lersek (re: dir_wrapper error handling)

(sorry for polluting this PR with this; this side discussion relates to https://github.com/scylladb/scylladb/pull/19624 instead)

@bhalevy, IIUC, you are hinting at patch#1, passing the error handler callback to the lambda that is chained on to the (possibly failed / failing) future, with `handle_exception`.

In my opinion, that's already being done in v4 of https://github.com/scylladb/scylladb/pull/19624. Namely, `sstable_write_io_check` calls `do_io_check`, which calls `futurize_invoke(...).handle_exception(...)` nearly identically to what is seen in patch#1. `futurize_invoke` is idempotent wrt. futures (i.e., it should have no effect if the `func` passed to it already returns a future), and then the lambda passed to `handle_exception` just calls `error_handler` -- same as in patch#1.

The only difference appears to be that patch#1 here prepares for `error_handler` itself throwing -- in that case, the lambda from patch#1 wraps that exception into the output future.

And, I'm not sure that's even necessary for patch#1 to do, because `handle_exception` in `seastar/include/seastar/core/future.hh` is documented as follows:

```
    /// When the future resolves, if it resolves with an exception,
    /// handle_exception(func) replaces the exception with the value
    /// returned by func. The exception is passed (as a std::exception_ptr)
    /// as a parameter to func; func may return the replacement value
    /// immediately (T or std::tuple<T...>) or in the future (future<T...>)
    /// and is even allowed to return (or throw) its own exception.
```

In other words, the extra care in the lamba of patch#1 seems redundant.

Anyway, I believe that either `sstable_write_io_check` or -- more frugally -- `do_io_check` should suffice for https://github.com/scylladb/scylladb/pull/19624. If your intent was to point out something else with the hint, please let me know; the above is how I can interpret your CC for now. Thanks!
